dnsprovider/aws.go

```diff
@@ -0,0 +1,315 @@
+		HostedZoneId: hostedZone.Id,
+	}
+
+	endpoints := []endpoint.Endpoint{}
+
+	f := func(resp *route53.ListResourceRecordSetsOutput, lastPage bool) (shouldContinue bool) {
+		for _, r := range resp.ResourceRecordSets {
+			if *r.Type != route53.RRTypeA {
+				continue
+			}
+
+			for _, rr := range r.ResourceRecords {
+				endpoint := endpoint.Endpoint{
+					DNSName: *r.Name,
+					Target:  *rr.Value,
+				}
+
+				endpoints = append(endpoints, endpoint)
+			}
+		}
+
+		return true
+	}
+
+	err = p.Client.ListResourceRecordSetsPages(params, f)
+	if err != nil {
+		return nil, err
+	}
+
+	return endpoints, nil
+}
+
+// CreateRecords creates a given set of DNS records in the given hosted zone.
+func (p *AWSProvider) CreateRecords(zone string, records []endpoint.Endpoint) error {
+	hostedZone, err := p.Zone(zone)
+	if err != nil {
+		return err
+	}
+
+	changes := []*route53.Change{}
+
+	for _, record := range records {
+		change := &route53.Change{
+			Action: aws.String(route53.ChangeActionCreate),
+			ResourceRecordSet: &route53.ResourceRecordSet{
+				Name: aws.String(record.DNSName),
+				ResourceRecords: []*route53.ResourceRecord{
+					{
+						Value: aws.String(record.Target),
+					},
+				},
+				TTL:  aws.Int64(300),
+				Type: aws.String(route53.RRTypeA),
+			},
+		}
+
+		changes = append(changes, change)
+	}
+
+	params := &route53.ChangeResourceRecordSetsInput{
+		HostedZoneId: hostedZone.Id,
+		ChangeBatch: &route53.ChangeBatch{
+			Changes: changes,
+		},
+	}
+
+	if p.DryRun {
+		log.Infof("Creating records: %#v", params.ChangeBatch.Changes)
+		return nil
+	}
+
+	_, err = p.Client.ChangeResourceRecordSets(params)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// UpdateRecords updates a given set of old records to a new set of records in a given hosted zone.
+func (p *AWSProvider) UpdateRecords(zone string, newRecords, _ []endpoint.Endpoint) error {
+	hostedZone, err := p.Zone(zone)
+	if err != nil {
+		return err
+	}
+
+	changes := []*route53.Change{}
+
+	for _, record := range newRecords {
+		change := &route53.Change{
+			Action: aws.String(route53.ChangeActionUpsert),
+			ResourceRecordSet: &route53.ResourceRecordSet{
+				Name: aws.String(record.DNSName),
+				ResourceRecords: []*route53.ResourceRecord{
+					{
+						Value: aws.String(record.Target),
+					},
+				},
+				TTL:  aws.Int64(300),
+				Type: aws.String(route53.RRTypeA),
+			},
+		}
+
+		changes = append(changes, change)
+	}
+
+	params := &route53.ChangeResourceRecordSetsInput{
+		HostedZoneId: hostedZone.Id,
+		ChangeBatch: &route53.ChangeBatch{
+			Changes: changes,
+		},
+	}
+
+	if p.DryRun {
+		log.Infof("Updating records: %#v", params.ChangeBatch.Changes)
+		return nil
+	}
+
+	_, err = p.Client.ChangeResourceRecordSets(params)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// DeleteRecords deletes a given set of DNS records in a given zone.
+func (p *AWSProvider) DeleteRecords(zone string, records []endpoint.Endpoint) error {
```

Comment:
what is the need of having three different methods for `create, update, delete` when all that is different is `Action` ?